app/ante/evm/04_validate.go

```diff
@@ -97,11 +97,23 @@ func ValidateTx(tx sdktypes.Tx) (*tx.Fee, error) {
 	return authInfo.Fee, nil
 }
 
-func CheckTxFee(txFeeInfo *tx.Fee, txFee sdktypes.Coins, txGasLimit uint64) error {
+func CheckTxFee(txFeeInfo *tx.Fee, txFee sdktypes.Coins, txGasLimit uint64, evmDenom string, denomDecimals uint32) error {
 	if txFeeInfo == nil {
 		return nil
 	}
 
+	// in case the evm has 6 decimals denom, we need to make the corresponding conversion
+	// before checking equality. The conversion rounds down, so we have to
+	// convert the txFee amt to 6 decimals
+	if denomDecimals == evmtypes.Denom6Dec {
+		for i := range txFee {
+			if txFee[i].Denom != evmDenom {
+				continue
+			}
+			txFee[i] = evmtypes.Convert18To6DecimalsCoin(txFee[i])
+		}
+	}
```

Comment:
why are we reducing the precision of txFee here instead of increasing the precision of txFeeInfo?

Reply:
because on the next line we're checking if they're equal. If we scale the txFeeInfo, we'll add some trailing zeros. And if we have non-zero decimals after the 6th decimal, the equality will never be true, even though these are discarded anyways